signalfx/resource_signalfx_detector_test.go

```diff
@@ -163,6 +164,7 @@ func TestAccCreateUpdateDetector(t *testing.T) {
 				Config: newDetectorConfig,
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckDetectorResourceExists,
+					waitBeforeTest,
```

Comment:
State is not re-read after this so this should not really do anything.

Maybe the [original issue with tags](https://github.com/splunk-terraform/terraform-provider-signalfx/pull/306#issuecomment-870417521) was resolved on the Splunk API side :thinking:

Reply:
I still got the error without sleep (here in tests or in the resource code as it was done in the original PR).

If I remove the sleep from this branch:

```
diff --git a/signalfx/resource_signalfx_detector_test.go b/signalfx/resource_signalfx_detector_test.go
index 3198bb9..8c1a571 100644
--- a/signalfx/resource_signalfx_detector_test.go
+++ b/signalfx/resource_signalfx_detector_test.go
@@ -164,7 +164,6 @@ func TestAccCreateUpdateDetector(t *testing.T) {
 				Config: newDetectorConfig,
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckDetectorResourceExists,
-					waitBeforeTest,
 					resource.TestCheckResourceAttr("signalfx_detector.application_delay", "name", "max average delay"),
 					resource.TestCheckResourceAttr("signalfx_detector.application_delay", "description", "your application is slow"),
 					resource.TestCheckResourceAttr("signalfx_detector.application_delay", "timezone", "Europe/Paris"),
```

here are the results with error:

```
make testacc TESTARGS='-run=TestAccCreateUpdateDetector -count=3'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccCreateUpdateDetector -count=3 -timeout 120m
?       github.com/splunk-terraform/terraform-provider-signalfx [no test files]
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (7.80s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (5.88s)
=== RUN   TestAccCreateUpdateDetector
    testing.go:684: Step 0 error: After applying this step and refreshing, the plan was not empty:
        
        DIFF:
        
        UPDATE: signalfx_detector.application_delay
          description:                  "your application is slow" => "your application is slow"
          disable_sampling:             "false" => "false"
          id:                           "E9ozxv5AAAA" => "E9ozxv5AAAA"
          max_delay:                    "30" => "30"
          min_delay:                    "15" => "15"
          name:                         "max average delay" => "max average delay"
          program_text:                 "signal = data('app.delay').max().publish('app delay')\ndetect(when(signal > 60, '5m')).publish('Processing old messages 5m')\ndetect(when(signal > 60, '30m')).publish('Processing old messages 30m')\n" => "signal = data('app.delay').max().publish('app delay')\ndetect(when(signal > 60, '5m')).publish('Processing old messages 5m')\ndetect(when(signal > 60, '30m')).publish('Processing old messages 30m')\n"
          rule.#:                       "2" => "2"
          rule.0.description:           "maximum > 60 for 30m" => "maximum > 60 for 30m"
          rule.0.detect_label:          "Processing old messages 30m" => "Processing old messages 30m"
          rule.0.disabled:              "false" => "false"
          rule.0.notifications.#:       "1" => "1"
          rule.0.notifications.0:       "Email,foo-alerts@example.com" => "Email,foo-alerts@example.com"
          rule.0.parameterized_body:    "" => ""
          rule.0.parameterized_subject: "" => ""
          rule.0.runbook_url:           "" => ""
          rule.0.severity:              "Critical" => "Critical"
          rule.0.tip:                   "" => ""
          rule.1.description:           "maximum > 60 for 5m" => "maximum > 60 for 5m"
          rule.1.detect_label:          "Processing old messages 5m" => "Processing old messages 5m"
          rule.1.disabled:              "false" => "false"
          rule.1.notifications.#:       "1" => "1"
          rule.1.notifications.0:       "Email,foo-alerts@example.com" => "Email,foo-alerts@example.com"
          rule.1.parameterized_body:    "" => ""
          rule.1.parameterized_subject: "" => ""
          rule.1.runbook_url:           "" => ""
          rule.1.severity:              "Warning" => "Warning"
          rule.1.tip:                   "" => ""
          show_data_markers:            "true" => "true"
          show_event_lines:             "false" => "false"
          tags.#:                       "0" => "2"
          tags.0:                       "" => "tag-1"
          tags.1:                       "" => "tag-2"
          teams.#:                      "1" => "1"
          teams.0:                      "E9oy16VAAAA" => "E9oy16VAAAA"
          time_range:                   "3600" => "3600"
          timezone:                     "Europe/Paris" => "Europe/Paris"
          url:                          "https://app.signalfx.com/#/detector/E9ozxv5AAAA" => "https://app.signalfx.com/#/detector/E9ozxv5AAAA"
          viz_options.#:                "1" => "1"
          viz_options.0.color:          "orange" => "orange"
          viz_options.0.display_name:   "" => ""
          viz_options.0.label:          "app delay" => "app delay"
          viz_options.0.value_prefix:   "" => ""
          viz_options.0.value_suffix:   "" => ""
          viz_options.0.value_unit:     "Second" => "Second"
        
        
        
        STATE:
        
        signalfx_detector.application_delay:
          ID = E9ozxv5AAAA
          provider = provider.signalfx
          description = your application is slow
          disable_sampling = false
          max_delay = 30
          min_delay = 15
          name = max average delay
          program_text = signal = data('app.delay').max().publish('app delay')
        detect(when(signal > 60, '5m')).publish('Processing old messages 5m')
        detect(when(signal > 60, '30m')).publish('Processing old messages 30m')
        
          rule.# = 2
          rule.0.description = maximum > 60 for 30m
          rule.0.detect_label = Processing old messages 30m
          rule.0.disabled = false
          rule.0.notifications.# = 1
          rule.0.notifications.0 = Email,foo-alerts@example.com
          rule.0.parameterized_body = 
          rule.0.parameterized_subject = 
          rule.0.runbook_url = 
          rule.0.severity = Critical
          rule.0.tip = 
          rule.1.description = maximum > 60 for 5m
          rule.1.detect_label = Processing old messages 5m
          rule.1.disabled = false
          rule.1.notifications.# = 1
          rule.1.notifications.0 = Email,foo-alerts@example.com
          rule.1.parameterized_body = 
          rule.1.parameterized_subject = 
          rule.1.runbook_url = 
          rule.1.severity = Warning
          rule.1.tip = 
          show_data_markers = true
          show_event_lines = false
          teams.# = 1
          teams.0 = E9oy16VAAAA
          time_range = 3600
          timezone = Europe/Paris
          url = https://app.signalfx.com/#/detector/E9ozxv5AAAA
          viz_options.# = 1
          viz_options.0.color = orange
          viz_options.0.display_name = 
          viz_options.0.label = app delay
          viz_options.0.value_prefix = 
          viz_options.0.value_suffix = 
          viz_options.0.value_unit = Second
        
          Dependencies:
            signalfx_team.detectorTeam
        signalfx_team.detectorTeam:
          ID = E9oy16VAAAA
          provider = provider.signalfx
          description = Detector Team
          name = Super Cool Team
          notifications_critical.# = 1
          notifications_critical.0 = Email,test@example.com
          notifications_default.# = 1
          notifications_default.0 = Webhook,,secret,https://www.example.com
          notifications_info.# = 1
          notifications_info.0 = Webhook,,secret,https://www.example.com/2
          notifications_major.# = 1
          notifications_major.0 = Webhook,,secret,https://www.example.com/3
          notifications_minor.# = 1
          notifications_minor.0 = Webhook,,secret,https://www.example.com/4
          notifications_warning.# = 1
          notifications_warning.0 = Webhook,,secret,https://www.example.com/5
          url = https://app.signalfx.com/#/team/E9oy16VAAAA
--- FAIL: TestAccCreateUpdateDetector (3.55s)
FAIL
FAIL    github.com/splunk-terraform/terraform-provider-signalfx/signalfx        17.252s
?       github.com/splunk-terraform/terraform-provider-signalfx/version [no test files]
FAIL
make: *** [GNUmakefile:17 : testacc] Erreur 1

```


whereas it works fine with this change:

```
make testacc TESTARGS='-run=TestAccCreateUpdateDetector -count=20'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccCreateUpdateDetector -count=20 -timeout 120m
?       github.com/splunk-terraform/terraform-provider-signalfx [no test files]
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.72s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.57s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.63s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.78s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (7.21s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.96s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.96s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.63s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.81s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.74s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.45s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.61s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.73s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (7.01s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (7.03s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.57s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.59s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.55s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.74s)
=== RUN   TestAccCreateUpdateDetector
--- PASS: TestAccCreateUpdateDetector (6.63s)
PASS
ok      github.com/splunk-terraform/terraform-provider-signalfx/signalfx        134.951s
?       github.com/splunk-terraform/terraform-provider-signalfx/version [no test files]

Reply:
Ok, this does work because it makes a pause inside the step 0 between the checks execution :
https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.15.0/helper/resource/testing_config.go#L109
and the plan/refresh ran just after:
https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.15.0/helper/resource/testing_config.go#L119
https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.15.0/helper/resource/testing_config.go#L127

To make this more obvious, I'd move this line after the last check of this step: https://github.com/splunk-terraform/terraform-provider-signalfx/pull/313/files#diff-9da637ecab2d74e99b4e9eb4d9dd6cbe61b3b922457cabc63942e714cae044a7R202

Reply:
should be good